independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/CurrentInterception.java

```diff
@@ -0,0 +1,35 @@
+package io.quarkus.arc.impl;
+
+import java.util.ArrayDeque;
+
+// this is used to prevent recursive interception
+// due to high cost and little usefulness, it is only used in the strict mode
+public class CurrentInterception {
+    private static final ThreadLocal<ArrayDeque<InterceptedMethodMetadata>> STACK = new ThreadLocal<>() {
+        @Override
+        protected ArrayDeque<InterceptedMethodMetadata> initialValue() {
+            return new ArrayDeque<>();
+        }
+    };
+
+    public static boolean mayIntercept(InterceptedMethodMetadata interception) {
+        ArrayDeque<InterceptedMethodMetadata> stack = STACK.get();
+        for (InterceptedMethodMetadata item : stack) {
```

Comment:
using streams would save you a few lines of code

Reply:
It would also add needless overhead. Thread locals are costly enough, I don't want to add more cost if I can easily avoid it.